
```{r}
#.libPaths('/data/gpfs/projects/punim2121/R_libs/4.1.2')
.libPaths('/data/gpfs/projects/punim2121/R_libs/4.2.1')
.libPaths()
```


```{r}
#BiocManager::install('scran')
#BiocManager::install("transformGamPoi")
```

```{r}
library(scran)
library(transformGamPoi)
library(SingleCellExperiment)
library(zellkonverter)
library(MatrixGenerics)
library(BiocParallel)
library(scDblFinder)
#library(usethis) 
#usethis::edit_r_environ()
```


## Read data after QC

```{r}
## Init list with reference genome names that were used for the 2 alignments
ref_genomes=c('GRCh38-p14-Gencode_v44','GRCh38-p13-Gencode_v33')

dat_list=list()
for (ref_genome in ref_genomes){

  
  #for (gene_tag in c('','_all_genes')){
  #fname=paste0('../../data/',ref_genome,'_data_QC',gene_tag,'.h5ad')
  fname=paste0('../../data/',ref_genome,'_data_QC.h5ad')
  dat=zellkonverter::readH5AD(fname,reader='python')
  print(ref_genome)
  #dat_list[paste0(ref_genome,gene_tag)]=dat
  
  #====== Plot mean-variance plot + fit curve with manually set Gamma-Poisson overdispersion factor ============
  ## Rename AssayName from 'X' to 'counts', as this is necessary for scDblFinder to run
  assayNames(dat)[grepl('cellbender',assayNames(dat))]='counts'
  
  sce=dat
  overdispersion=5
  # Exclude genes where all counts are zero
  #sce <- sce[rowMeans2(counts(sce)) > 0, ]
  
  gene_means <- rowMeans2(counts(sce))
  gene_var <- rowVars(counts(sce))
  plot(gene_means, gene_var, log = "xy", main=paste0(ref_genome," - Log-log scatter plot of mean vs variance"))
  abline(a = 0, b = 1)
  sorted_means <- sort(gene_means)
  lines(sorted_means, sorted_means + overdispersion * sorted_means^2, col = "purple")
  
  
  
  #====== Calculate size factors for all cells with scran -> use Fastcluster clusters for this ===========
  sce$fastcluster<-fastcluster(sce, iter.max=100)
  input_groups=sce@colData[,'fastcluster',drop=T]
  data_mat=((assay(sce)))
  register(MulticoreParam(workers = 4))
  size_factors = calculateSumFactors(data_mat, clusters=input_groups)
  rm(data_mat)
  
  
  
  #====== Apply acosh and (shifted log) transformation to the raw counts ============
  ## If memory is enough ~ >64GB, on_disk can be set to FALSE -> calculation is much faster using memory
  ## If memory is sparse, set on_disk=TRUE, this way there is no out-of-memory error, in exchange for slower speed
  assay(sce, withDimnames=FALSE,"acosh")=acosh_transform(assay(sce, "counts"),size_factors=size_factors,overdispersion=TRUE,on_disk = FALSE)
  #assay(sce, withDimnames=FALSE,"shifted_log")=shifted_log_transform(assay(sce,"counts"),size_factors=size_factors,overdispersion=TRUE,on_disk=TRUE)
  
  ## Plot mean-variance plot of normalised data
  acosh_var <- rowVars(assay(sce, "acosh"))
  plot(gene_means, acosh_var, log = "x", main = "Log expression vs variance of acosh stabilized values")
  abline(h = 1)
  
  #shift_log_var <- rowVars(assay(sce, "shifted_log"))
  #plot(gene_means, shift_log_var, log = "x", main = "Log expression vs variance of shifted log stabilized values")
  #abline(h = 1)
  
  
  #========== Save file ===========
  ## Rename AssayName from 'X' to 'counts', as this is necessary for scDblFinder to run
  assayNames(sce)[grepl('counts',assayNames(sce))]='cellbender'
  
  #fn2=paste0('../../data/',ref_genome,'_data_QC_norm',gene_tag,'.h5ad')
  fn2=paste0('../../data/',ref_genome,'_data_QC_norm.h5ad')
  writeH5AD(sce,file=fn2,compression='gzip')
  #}
}
```



